sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDB.scala

```diff
@@ -0,0 +1,453 @@
+        logWarning(s"Error recursively deleting local dir $file while $msg", e)
+    }
+  }
+
+  /** Records the duration of running `body` for the next query progress update. */
+  protected def timeTakenMs(body: => Unit): Long = Utils.timeTakenMs(body)._2
+
+  override protected def logName: String = s"${super.logName} $loggingId"
+}
+
+
+/** Mutable and reusable pair of byte arrays */
+class ByteArrayPair(var key: Array[Byte] = null, var value: Array[Byte] = null) {
+  def set(key: Array[Byte], value: Array[Byte]): ByteArrayPair = {
+    this.key = key
+    this.value = value
+    this
+  }
+}
+
+
+/**
+ * Configurations for optimizing RocksDB
+ *
+ * @param compactOnCommit Whether to compact RocksDB data before commit / checkpointing
+ */
+case class RocksDBConf(
+    minVersionsToRetain: Int,
+    compactOnCommit: Boolean,
+    pauseBackgroundWorkForCommit: Boolean,
+    blockSizeKB: Long,
+    blockCacheSizeMB: Long,
+    lockAcquireTimeoutMs: Long)
+
+object RocksDBConf {
+  /** Common prefix of all confs in SQLConf that affects RocksDB */
+  val ROCKSDB_CONF_NAME_PREFIX = "spark.sql.streaming.stateStore.rocksdb"
+
+  private case class ConfEntry(name: String, default: String) {
+    def fullName: String = s"$ROCKSDB_CONF_NAME_PREFIX.${name}".toLowerCase(Locale.ROOT)
+  }
+
+  // Configuration that specifies whether to compact the RocksDB data every time data is committed
+  private val COMPACT_ON_COMMIT_CONF = ConfEntry("compactOnCommit", "false")
+  private val PAUSE_BG_WORK_FOR_COMMIT_CONF = ConfEntry("pauseBackgroundWorkForCommit", "true")
+  private val BLOCK_SIZE_KB_CONF = ConfEntry("blockSizeKB", "4")
+  private val BLOCK_CACHE_SIZE_MB_CONF = ConfEntry("blockCacheSizeMB", "8")
+  private val LOCK_ACQUIRE_TIMEOUT_MS_CONF = ConfEntry("lockAcquireTimeoutMs", "60000")
+
+  def apply(storeConf: StateStoreConf): RocksDBConf = {
+    val confs = CaseInsensitiveMap[String](storeConf.confs)
+
+    def getBooleanConf(conf: ConfEntry): Boolean = {
+      Try { confs.getOrElse(conf.fullName, conf.default).toBoolean } getOrElse {
+        throw new IllegalArgumentException(s"Invalid value for '${conf.fullName}', must be boolean")
+      }
+    }
+
+    def getPositiveLongConf(conf: ConfEntry): Long = {
+      Try { confs.getOrElse(conf.fullName, conf.default).toLong } filter { _ >= 0 } getOrElse {
+        throw new IllegalArgumentException(
+          s"Invalid value for '${conf.fullName}', must be a positive integer")
+      }
+    }
+
+    RocksDBConf(
+      storeConf.minVersionsToRetain,
+      getBooleanConf(COMPACT_ON_COMMIT_CONF),
+      getBooleanConf(PAUSE_BG_WORK_FOR_COMMIT_CONF),
+      getPositiveLongConf(BLOCK_SIZE_KB_CONF),
+      getPositiveLongConf(BLOCK_CACHE_SIZE_MB_CONF),
+      getPositiveLongConf(LOCK_ACQUIRE_TIMEOUT_MS_CONF))
+  }
+
+  def apply(): RocksDBConf = apply(new StateStoreConf())
+}
+
+case class AcquiredThreadInfo(
+    threadRef: WeakReference[Thread] = new WeakReference[Thread](Thread.currentThread()),
```

Comment:
Is there any caller providing these parameters manually?

Reply:
No. Make `threadRef` and `tc` as public member in the next commit.